Retain hook does not support ndarray

Process of exporting retains references of input value to output temporary values using `RetainInputHook`. This hook only supports `chainer.Variable`, so if the target model handles `chainer.Chain` accepting `numpy.ndarray`, cause ValueError.

```python
class Model(chainer.Chain):

    def forward(self, x, h):
        return x + h

class ParentModel(chainer.Chain):

    def __init__(self):
        super().__init__()
        self.h = np.array(5, dtype=np.float32)
        with self.init_scope():
            self.m = Model()

    def forward(self, x):
        return self.m(x, self.h)

x  = chainer.Variable(np.array(3, dtype=np.float32))
export(ParentModel(), (x,))
```

```
E           ValueError: type <class 'numpy.ndarray'> is not supported to retain input
```

ref #201

--- a/onnx_chainer/export.py
+++ b/onnx_chainer/export.py
@@ -186,6 +186,9 @@ class RetainInputHook(chainer.LinkHook):
         self.hooked_apply = hooked_apply
 
     def _extract_inputs(self, args):
+        # Retain only chainer.Variable (and its collection)
+        # Other type args are ignored and not checked instance IDs
+        # If these variable are used in FunctionNode, they will be retained
         ret = set()
         if isinstance(args, chainer.Variable):
             ret.add(id(args))
@@ -195,9 +198,6 @@ class RetainInputHook(chainer.LinkHook):
         elif isinstance(args, dict):
             for arg in args.values():
                 ret |= self._extract_inputs(arg)
-        else:
-            raise ValueError('type {} is not supported to retain input'.format(
-                type(args)))
         return ret
 
     def forward_preprocess(self, args):

--- a/tests/test_inout.py
+++ b/tests/test_inout.py
@@ -142,7 +142,7 @@ class TestRetainInputHook(object):
 
     @pytest.mark.parametrize(
         'test_type', ['variable', 'list', 'dict', 'array'])
-    def test_hook(self, test_type):
+    def test_hook_for_funcnode(self, test_type):
         class Model(chainer.Chain):
 
             def forward(self, x):
@@ -162,6 +162,37 @@ class TestRetainInputHook(object):
             # input is ndarray and not checked in forward_preprocess
             expected_count += 1
         assert len(h.retain_inputs) == expected_count
+
+    @pytest.mark.parametrize('test_type', ['array'])
+    def test_hook_for_childlink(self, test_type):
+        # TODO(disktnk): test_type='variable' is failed
+        class ChildModel(chainer.Chain):
+
+            def forward(self, x, h):
+                if test_type in ['variable', 'array']:
+                    h = [chainer.as_variable(h)]
+                elif test_type == 'dict':
+                    h = list(h.values())
+                h.append(x)
+                return F.stack(h)
+
+        class ParentModel(chainer.Chain):
+
+            def __init__(self, get_x):
+                super().__init__()
+                self.get_x = get_x
+                with self.init_scope():
+                    self.m = ChildModel()
+
+            def forward(self, x):
+                h = self.get_x(test_type)
+                return self.m(x, h)
+
+        model = ParentModel(self.get_x)
+        x = self.get_x('variable')
+        with RetainInputHook() as h:
+            model(x)
+        assert len(h.retain_inputs) == 1
 
 
 @testing.parameterize(

--- a/tests/test_inout.py
+++ b/tests/test_inout.py
@@ -134,17 +134,20 @@ class TestRetainInputHook(object):
                 chainer.Variable(np.array(5, dtype=np.float32))]
         elif test_type == 'dict':
             return {'x': chainer.Variable(np.array(3, dtype=np.float32))}
+        elif test_type == 'array':
+            return np.array(3, dtype=np.float32)
         else:
             assert test_type == 'variable'
             return chainer.Variable(np.array(3, dtype=np.float32))
 
-    @pytest.mark.parametrize('test_type', ['variable', 'list', 'dict'])
+    @pytest.mark.parametrize(
+        'test_type', ['variable', 'list', 'dict', 'array'])
     def test_hook(self, test_type):
         class Model(chainer.Chain):
 
             def forward(self, x):
-                if test_type == 'variable':
-                    x = [x]
+                if test_type in ['variable', 'array']:
+                    x = [chainer.as_variable(x)]
                 elif test_type == 'dict':
                     x = list(x.values())
                 x.append(chainer.Variable(np.array(7, np.float32)))
@@ -154,7 +157,11 @@ class TestRetainInputHook(object):
         x = self.get_x(test_type)
         with RetainInputHook() as h:
             model(x)
-        assert len(h.retain_inputs) == 1
+        expected_count = 1
+        if test_type == 'array':
+            # input is ndarray and not checked in forward_preprocess
+            expected_count += 1
+        assert len(h.retain_inputs) == expected_count
 
 
 @testing.parameterize(

